### ISSUE TITLE
Docker: Add release type on docker tags

### ISSUE THREAD

**Issue by stavros-k** [storj/storj#5422]
**Describe the feature or problem you'd like to solve**

Currently scheme of docker tags are not very friendly to allow downstream projects to update the container to the latest stable tag using automated tools, such as renovate or even custom scripts.

**Propose a Solution**

Add a prefix, suffix or anywhere in the middle, the release type. eg

For a stable release
`1d42f9ac3-v1.68.2-go1.18.8-stable`

For a pre-release
`sometihng-v1.66.4-go1.18.8-prerelease`

With the above one can easily use regex matching.

Thanks

**Additional context**

**Comment by ifraixedes** [comment]
@stavros-k thank you for your suggestion.

Can you please tell us if you are referring to all the images (satellite, storagenode, etc.) or to only a specific one?

**Comment by stavros-k** [comment]
> @stavros-k thank you for your suggestion.
> 
> 
> 
> Can you please tell us if you are referring to all the images (satellite, storagenode, etc.) or to only a specific one?

Right. Sorry!

I'm interested in storagenode.
But ideally something like that would be beneficial to all of them

**Comment by ifraixedes** [comment]
We don't rollout pre-release for the Storagenode, we only keep it in pre-release when we start to rollout the new version, but once we publish the [100% cursor for the version]( https://version.storj.io/), we tag the Docker image as `latest` between others. My understanding is that you want to try these pre-relase images for some reason because if you want to use the last stable release image, you can pull the `latest` tag.

Nonetheless, this is going to change because right now the Docker images don't contain the binaries and they download it when they start, always picking the version that matches the cursor, so we are going to stop to publish Storagenode Docker images for each release. You can follow that change at https://github.com/storj/storagenode-docker/issues/1

**Comment by stavros-k** [comment]
No, I don't want to try pre-releases. I want to **only** pull the stable versions.
For example 1.66.4 is marked as pre-release in github, yet it has a container on dockerhub.

Using latest tag is a very bad practice in containerized environments. 
Same goes for download/installing/building at runtime.

Feel free to close this issue

**Comment by Ornias1993** [comment]
To chime in on this:
There are two things at play here:
1. Using clearly defined versions
This is important to ensure the user (and potential downstreams!) know what they are going to be running and, more importantly, allows automated updating software to detect when a new version number is released and, in a perfect world, check if it's a Major, Patch or Minor release.

2. Using repeatable builds
One of the reasons the world is moving to containerised deployments is repeatable builds. Fetching binaries on container start, is an absolutely ridiculous idea outside of a low-key home-user environment. Generally speaking, users should even use digest pinning more, to ensure repeatable builds... Not get images that, inherently, make it harder to make repeatable builds.

That's not to say that fetching binaries is always bad, for example Nextcloud managed to implement repeatable container builds with fetching the right version at container start just fine. Because they sync the container tag, with the build they are going to fetch, simply put: The administrator knows 100% what he can expect when the software is deployed.

**Comment by ifraixedes** [comment]
> Using latest tag is a very bad practice in containerized environments.

I agree and probably quite a few of my colleagues too.

> Same goes for download/installing/building at runtime.

I also agree, but there were other things to consider aside from technical aspects.

> Basically, what you're saying is that you now have undefined container versions and are moving to also make the content of the container be fetched at start. 

Not moving, it's in that way for several months ago.

> I honestly, from a professional perspective, have no idea which containerisation expert has come up with that plan, because it simply does not make any sense.

As commented above, there were other non-tech aspects to consider.

I'm not a decision-maker here and I agree with all of those technical aspects. The only solution that I see is having "normal" images (the ones that don't mutate) with the version as of now but adding some postfix for distinguishing between stable and other kinds of releases and keeping "abnormal" images. That would work for everybody.

I'm fine to raise a conversation about these in our next general engineering meeting, I guess that the trade-off is to make our release process more complex and everything that implies, but worth to discuss it.

I'm keeping the issue open until we have that meeting.

**Comment by Ornias1993** [comment]
> > Using latest tag is a very bad practice in containerized environments.
> 
> I agree and probably quite a few of my colleagues too.
> 
> > Same goes for download/installing/building at runtime.
> 
> I also agree, but there were other things to consider aside from technical aspects.
> 
> > Basically, what you're saying is that you now have undefined container versions and are moving to also make the content of the container be fetched at start.
> 
> Not moving, it's in that way for several months ago.
> 
> > I honestly, from a professional perspective, have no idea which containerisation expert has come up with that plan, because it simply does not make any sense.
> 
> As commented above, there were other non-tech aspects to consider.
> 
> I'm not a decision-maker here and I agree with all of those technical aspects. The only solution that I see is having "normal" images (the ones that don't mutate) with the version as of now but adding some postfix for distinguishing between stable and other kinds of releases and keeping "abnormal" images. That would work for everybody.
> 
> I'm fine to raise a conversation about these in our next general engineering meeting, I guess that the trade-off is to make our release process more complex and everything that implies, but worth to discuss it.
> 
> I'm keeping the issue open until we have that meeting.

Thanks for the thorough response, it's appreciated :)

For as a downstream offering it in our App catalog (TrueCharts), having clear version tags and stable images (what's inside them won't change), is the difference between: "We tested this version and it works" and "We cannot actually test this, because it might randomly change, use at your own risk".

I don't think that's the experience or impression storj wants to offer...

I get why one would want to do the "fetch latest on start" though, because it guarantees somewhat that all users are actually using the latest software and update with at least downtime as possible... So just, stabilising images would take some design work and I definately won't expect such a thing in a week ;-)

**Comment by ifraixedes** [comment]
> Thanks for the thorough response, it's appreciated :)

To you for providing this feedback, we appreciate it.

> For as a downstream offering it in our App catalog (TrueCharts), having clear version tags and stable images (what's inside them won't change), is the difference between: "We tested this version and it works" and "We cannot actually test this, because it might randomly change, use at your own risk".

I totally understand this case and I hope that most of my colleagues, if not all, will.

> I don't think that's the experience or impression storj wants to offer...

For sure no. If I remember well, we needed to have those images because a big part of potential operators wished to run a node without requiring almost technical expertise and with less maintenance as possible and they wanted to run it in Docker.

> I get why one would want to do the "fetch latest on start" though, because it guarantees somewhat that all users are actually using the latest software and update with at least downtime as possible... So just, stabilising images would take some design work and I definately won't expect such a thing in a week ;-)

It's more about that if a node is behind the ["minimum" version ](https://version.storj.io/), it stops receiving uploads; combining this with what I commented just above and with the issues that we had with [watchtower](https://github.com/storj/watchtower), we ended up with that solution. 

Thank you for your comprehension that if we decide to publish again "normal" images, it won't be immediate, not only for design work but also because we have a roadmap and we have to prioritize tasks.
